torrent.go

```diff
@@ -122,6 +123,44 @@ func (t *Torrent) Closed() <-chan struct{} {
 	return t.closed.LockedChan(&t.cl.mu)
 }
 
+// KnownSwarm returns the known subset of the peers in the Torrent's swarm, including active,
+// pending, and half-open peers.
+func (t *Torrent) KnownSwarm() (ks []Peer) {
+	// Add pending peers to the list
+	for _, peer := range t.peers {
+		ks = append(ks, peer)
+	}
+
+	// Add active peers to the list
+	for conn := range t.conns {
+		host, portString, err := net.SplitHostPort(conn.remoteAddr().String())
+		if err != nil {
+			panic(err)
+		}
+
+		ip := net.ParseIP(host)
+		port, err := strconv.Atoi(portString)
+		if err != nil {
+			panic(err)
+		}
+
+		ks = append(ks, Peer{
+			Id: conn.PeerID,
+			IP: ip,
+			Port: port,
+			Source: conn.Discovery,
+			// TODO: the connection can be unencrypted due to our (or the peer's) preference,
+			//       but the remote peer might support the encryption. Find a better way to query
+			//       that information, if possible.
+			SupportsEncryption: conn.encrypted,
+		})
+	}
+
+	// TODO: how can we add half-open peers?
```

Comment:
You can get the addresses from Torrent.halfOpen.

Reply:
But how do I get their peer ID, *source*, and whether or not they support encryption? `Torrent.halfOpen` just have the hostport strings right?

Reply:
That's right. I don't think we can as it is. We could however carry over the values when they're pulled from the pending peers list, and put it in the half-open map. Once the connection is successful, we can apply them there.

Reply:
See b39df82fbc9269c69257ac22dea87338a886faf6